ENT-5750 Keep killed flows that were started with client ids

lankydan opened the pull request:
If a flow is started with a client id, do not delete it from the
database. Instead, set the status to `KILLED` and store a
`KilledFlowException` in the database.

Keeping it around allows it to be reattached to using the same
mechanisms as active, completed or failed flows. Furthermore, without
this change it could be possible for a flow to be triggered again if
killed while the reconnecting rpc client is being used.

If there is no client id related to the flow, then kill flow keeps its
original behaviour of deleting all traces of the flow.

Flows cannot be killed if they are `COMPLETED`, `FAILED` or `KILLED`
already.

Logs have been added if requests to kill flows in these statuses are
made.

Do not update the status + persist the exception if the flow was killed
during flow initialisation (before persisting its first checkpoint).

Remove the client id mapping if the flow was killed and did not persist
its original checkpoint.

lankydan committed:
ENT-5750 Keep killed flows that were started with client ids

If a flow is started with a client id, do not delete it from the
database. Instead, set the status to `KILLED` and store a
`KilledFlowException` in the database.

Keeping it around allows it to be reattached to using the same
mechanisms as active, completed or failed flows. Furthermore, without
this change it could be possible for a flow to be triggered again if
killed while the reconnecting rpc client is being used.

If there is no client id related to the flow, then kill flow keeps its
original behaviour of deleting all traces of the flow.

lankydan committed:
ENT-5750 detekt

rick-r3 commented on node/src/integration-test/kotlin/net/corda/node/flows/KillFlowTest.kt:
```diff
@@ -210,6 +211,26 @@ class KillFlowTest {
+                    }
+                }
+                assertTrue(time < 1.minutes.toMillis(), "It should at a minimum, take less than a minute to kill this flow")
+                assertTrue(time < 5.seconds.toMillis(), "Really, it should take less than a few seconds to kill a flow")
```
I'm not sure you'll get away with this in CI.  I would remove it.  All it takes is a big GC or something.

lankydan replied:
Ive got it on all my other tests in this class. I can remove them.

kyriathar replied:
If we are not removing it after all, i guess then the -time < 1 min assertion- is redundant? We should consolidate the two?

rick-r3 reviewed (commented):
Interesting question.  What does kill flow on a KILLED flow do?

lankydan committed:
ENT-5750 un used veriable

kyriathar commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -355,10 +357,18 @@ internal class SingleThreadedStateMachineManager(
-                    checkpointStorage.removeCheckpoint(id, mayHavePersistentResults = true)
+                    if (flow.fiber.clientId != null) {
+                        checkpointStorage.updateStatus(id, Checkpoint.FlowStatus.KILLED)
+                        checkpointStorage.removeFlowException(id)
```
Should we check previous `Checkpoint.FlowStatus` to determine whether or not there is a database exception to remove it? I guess kill flow requests are going to be rare so it shouldn't make much difference though.

kyriathar commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowClientIdTests.kt:
```diff
@@ -727,35 +773,113 @@ class FlowClientIdTests {
+    private fun TestStartedNode.waitForOvernightObservation(id: StateMachineRunId, timeout: Duration) {
+        val timeoutTime = Instant.now().plusSeconds(timeout.seconds)
+        var exists = false
+        while (Instant.now().isBefore(timeoutTime) && !exists) {
```
Is this a busy waiting kind of? Should we add a `Thread.sleep` somewhere in its block?

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/flows/KillFlowTest.kt:
```diff
@@ -482,6 +503,15 @@ class KillFlowTest {
 
+    @StartableByRPC
+    @InitiatingFlow
+    class AFlowThatGetsMurderedWhileInTheHospital : FlowLogic<Unit>() {
```
Pernickety: this is a flow that gets hospitalized only, name is a bit implicit to what will happen to it.

kyriathar commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -600,14 +600,30 @@ class DBCheckpointStorage(
         return serializedFlowException?.deserialize(context = SerializationDefaults.STORAGE_CONTEXT)
     }
 
+    override fun addFlowException(id: StateMachineRunId, exception: Throwable) {
```
Would it be better if  we consolidate this method with `createDBFlowException` if we take `errorState.errors.last().exception.let {` outside?

kyriathar commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -600,14 +600,30 @@ class DBCheckpointStorage(
+                message = exception.message?.truncate(MAX_EXC_MSG_LENGTH, false),
+                stackTrace = exception.stackTraceToString(),
+                value = exception.storageSerialize().bytes,
+                persistedInstant = Instant.now(clock)
```
Could we use `clock.instant()` for consistency with other places in this class?

kyriathar commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -600,14 +600,30 @@ class DBCheckpointStorage(
+        currentDBSession()
+            .createNativeQuery("Update ${NODE_DATABASE_PREFIX}checkpoints set status = :status, timestamp = :timestamp where flow_id = :id")
+            .setParameter("status", flowStatus.ordinal)
+            .setParameter("timestamp", Instant.now(clock))
```
Same as above.

lankydan committed:
ENT-5750 Cannot kill finished flows

Flows cannot be killed if they are `COMPLETED`, `FAILED` or `KILLED`
already.

Logs have been added if requests to kill flows in these statuses are
made.

Do not update the status + persist the exception if the flow was killed
during flow initialisation (before persisting its first checkpoint).

Remove the client id mapping if the flow was killed and did not persist
its original checkpoint.

rick-r3 reviewed (dismissed):
Looks good from my perspective.  Up to you guys whether you make the changes @kyriathar suggested but I think we have an RC cutting today, so could do a follow up for expediency.

rick-r3 commented:
Needs some updates anyway, so might be time to fix some of the suggestions.

kyriathar commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowClientIdTests.kt:
```diff
@@ -308,6 +306,32 @@ class FlowClientIdTests {
+        assertTrue(aliceNode.hasStatus(flowHandle0!!.id, Checkpoint.FlowStatus.KILLED))
+        assertTrue(aliceNode.hasException(flowHandle0!!.id))
+
+        aliceNode.internals.smm.killFlow(flowHandle0!!.id)
```
Minor one but could we also assert that this `killFlow` returns a `false`?

lankydan committed:
ENT-5750 killed flow retained after node restart

lankydan committed:
ENT-5750 tidy up

lankydan committed:
ENT-5750 tidy up

lankydan commented:
Run tests: slow

lankydan committed:
ENT-5750 ...

lankydan committed:
ENT-5750 ...
